level/level.go

```diff
@@ -96,6 +101,35 @@ func AllowNone() Option {
 	return allowed(0)
 }
 
+// ParseOption creates an Option from a string value. It might be useful in case we need to configure
+// the log level using a flag, an environment variable, a configuration file, ...
+//
+// Leading and trailing spaces are ignored and the comparison is Case Insensitive:
+// "ERROR", "Error", " Error " are accepted
+//
+// If the level parameter does not match any of the available values, the function returns a not nil error.
+func ParseOption(level string) (Option, error) {
+	level = strings.TrimSpace(strings.ToLower(level))
+
+	switch level {
+	case debugValue.name:
+		return AllowDebug(), nil
+
+	case infoValue.name:
+		return AllowInfo(), nil
+
+	case warnValue.name:
+		return AllowWarn(), nil
+
+	case errorValue.name:
+		return AllowError(), nil
+
+	default:
+		// return a noop Option in case of unknown level string
+		return nil, fmt.Errorf("invalid level: %s", level)
+	}
+}
```

Comment:
```suggestion
// ParseLevel parses a string into its corresponding log level. Valid levels are
// "debug", "info", "warn", and "error". Levels are normalized via
// strings.TrimSpace and strings.ToLower.
func ParseLevel(level string) (Option, error) {
	switch strings.TrimSpace(strings.ToLower(level)) {
	case debugValue.name:
		return AllowDebug(), nil
	case infoValue.name:
		return AllowInfo(), nil
	case warnValue.name:
		return AllowWarn(), nil
	case errorValue.name:
		return AllowError(), nil
	default:
		return nil, fmt.Errorf("invalid level: %s", level)
	}
}

// MustParseLevel calls ParseLevel and panicks on error.
func MustParseLevel(level string) Option {
	opt, err := ParseLeveL(level)
	if err != nil {
		panic(err)
	}
	return opt
}
```

Reply:
Not totally confident on `ParseLevel` — but pretty sure `ParseOption` isn't right, because Options aren't necessarily defining levels, right? Eager for feedback there. Also, should we have an additional version of the function that takes a default value in case of error?

Reply:
> `logger = level.NewFilter(logger, level.AllowInfo(), level.Allow(level.MustParse("info")))`

I'm dubious of giving NewFilter multiple options which manipulate the same property. I'd prefer the allow level to be clearly established, and then that explicit level provided to the constructor. Ideally that first step could be accomplished in a single expression, for all common use cases. So... maybe... 

```go
var logger log.Logger
logger = log.NewLogfmtLogger(os.Stdout)
logger = level.NewFilter(logger, level.ParseDefault(*lv, "info"))
```

Implying my `ParseLevel` should be just `Parse`?

Reply:
> I'm dubious of giving NewFilter multiple options which manipulate the same property.

I find it nicer, although I admit it requires more insights into how functional options work.

> `level.ParseDefault(*lv, "info")`

Why not use a constant as a second parameter? Makes the API safer to use.

`level.Allow(level.ParseDefault(*lv, "info"))` still feels more "correct" from a type perspective, even if it's more verbose.

Reply:
> `level.Allow(level.ParseDefault(*lv, "info"))` still feels more "correct" from a type perspective, even if it's more verbose.

This seems good to me 👍 But the only difference seems to be the return value of ParseDefault?

Reply:
Correct. IMO it should return a Value, not an Option. Also, ParseDefault should accept a Value as well.

Can't really reason why, but this API just feels more natural and correct.

Reply:
> ParseDefault should accept a Value

Don't ParseXxx functions usually accept strings and produce domain entities? 🤔

Reply:
I'm talking about the second (fallback) parameter. The first one is what we are really parsing, isn't it? The second one is a fallback value that should always be "valid" (unless I'm mistaken what that parameter is).

Reply:
> I'm talking about the second (fallback) parameter. The first one is what we are really parsing, isn't it? 

Ah, so if you're suggesting

```go
func ParseDefault(s string, def Option) Option { ... }
```

and used as

```go
lv := level.ParseDefault(*lv, level.AllowInfo())
```

then that makes sense to me 👍

Reply:
I'm a bit confused now.
not sure to understand what will be an acceptable API. I'm open to implement the proposed API but looks like we don't have an agreement on which API will be the "accepted" one.

shall we move o a DIscussion and come back to a PR only after we found an agreement ?

Reply:
> Well, I'm suggesting
> 
> ```go
> func ParseDefault(s string, def Value) Value { ... }
> ```
> 
> used as
> 
> ```go
> opt := level.Allow(level.ParseDefault(*lv, level.InfoValue()))
> ```

`level.Allow` doesn't exist right now, so you're suggesting to create that as well?

Reply:
Apparently, I do (didn't know it didn't exist before).

An alternative I can imagine (if we want to avoid creating `level.Allow`) is something like this:

```go
func ParseAndAllow(s string, def Value) Option { ... }
```

If we see little value in having separate Parse and Allow functions, then something like this could be a compromise to directly create an Option from a parsed value while keeping the right semantics.

Reply:
So

> Parsing a level (which is represented by Value) makes more sense to me than parsing an option.

In the abstract I agree 100%, but NewFilter — which is the primary way consumers use the package — takes options, not values. I'm not against adding some parsing functionality that yields values, but I think we gotta have something that yields values which can be given directly to the NewFilter constructor.

Reply:
Yeah, that's a valid point, hence the `ParseAndAllow` suggestion. If we go that way, I'm not even sure we need to create `Allow` (or `Parse/ParseDefault`) now, but I'd keep the option open to create a `Parse` function later that yields values.

So semantically, we are looking for a function name that hints both parsing and allowing a level:

```go
func ParseAndAllow(s string, def Value) Option { ... }
```

I'm not completely sold on the name, but I believe we are looking for something like this.

Reply:
What's the difference between `def Value` and `def Option`?

Reply:
Hm, I guess not much if the returned value is an `Option`.

```go
logger = level.NewFilter(logger, ParseAndAllow(*lv, level.InfoValue()))
```

vs


```go
logger = level.NewFilter(logger, ParseAndAllow(*lv, level.AllowInfo()))
```

I doubt the end user would care, but it might be slightly easier to understand that when parsing an arbitrary string fails, it falls back to a _value_, not an _option_. But again, I doubt it matters much, both sound acceptable.

Reply:
FWIW, in https://github.com/go-kit/log/pull/22#discussion_r839064447 I suggested an API similar to what @sagikazarmark suggests above.

>Are you suggesting `func Parse(level string) (Option, error)` or `func Parse(level string) (Value, error)`? The second would mesh with something like `func Allow(v Value) Option` nicely. That new option could also be used like `level.Allow(level.Info())`.

At this point I feel like we we missed an opportunity for the `level.Allow(level.Info())` approach in our original design. Had we taken that approach originally we would have a more orthogonal and slimmer package API and it would make the current discussion easier to conclude.

Reply:
Hello,

my 2 cents, in the function with default I'd use an `Option` instead of a `Value` because this function is supposed to return an `Option` and then it's clear that the default parameter is the option returned in the parsing fails.

I'll try to summarize; looks like we wanna have the following API
```go
func ParseOption(level string) (Option, error)

func MustParseOption(level string) Option)

func ParseOptionWithDefault(level string, default Option) Option
```

If we think `ParseOption` is too verbose we could use `Parse`.

Any thought ?

Reply:
> Had we taken that approach originally we would have a more orthogonal and slimmer package API and it would make the current discussion easier to conclude.

@ChrisHines  We can always introduce that new API deprecating (but leaving in place) the old one. WDYT about opening a separate discussion about these API changes?

> If we think `ParseOption` is too verbose we could use `Parse`.

👎 for `Parse` returning an option. It'd lead to a more inconsistent API in my opinion.

Reply:
@sagikazarmark,  if `Parse` returns a `Value`, how to use with `NewFilter that accepts only options ?

Reply:
@mcosta74  See the thread above for examples.

Reply:
@sagikazarmark 

> 👎 for Parse returning an option. It'd lead to a more inconsistent API in my opinion.

I get the rationale here, but I'm afraid it hurts usability. The current package API only uses the Value type in the NewInjector function, which is pretty niche. The main entry point for users is the NewFilter constructor, which, for better or worse, takes Options and not Values.

I'd hope that whatever we add in this PR could be used directly by that constructor, which I guess means it would have to return an Option. But maybe there are other approaches? The goal, ultimately, is to enable the common use case of decorating a log.Logger with a level.Filter based on user input, without requiring user code to parse strings or convert types or whatever.

The simplest form of that I can think of would be

```go
var logger log.Logger
logger = log.NewLogfmtLogger(os.Stdout)
logger = level.NewFilter(logger, level.???(userString, optionalDefaultValue)) 
```

Which implies the new function should return an Option. Is there a way to get that kind of user experience with Parse functions that return Values?

Reply:
@peterbourgon I think we agree on both the desired user experience and the API necessary to achieve that, so this mostly comes down to a naming question.

I suggested `ParseAndAllow` to keep our options open for a future `Parse` function returning a value.
Some alternatives:
- `ParseAllow`
- `AllowFromString`
- `ParseAllowOption`

If we ever decide to make the changes we discussed to the rest of the API (`level.Allow(level.Info())`), `Parse` will come in handy.

Reply:
> The simplest form of that I can think of would be
> 
> ```go
> var logger log.Logger
> logger = log.NewLogfmtLogger(os.Stdout)
> logger = level.NewFilter(logger, level.???(userString, optionalDefaultValue)) 
> ```
> 
> Which implies the new function should return an Option. Is there a way to get that kind of user experience with Parse functions that return Values?

I see that more as the minimum set of functions added to the current API. That's not necessarily the simplest API overall.

> The current package API only uses the Value type in the NewInjector function, which is pretty niche.

Perhaps using the Value types aren't used enough, then.

-----------

All of that said, this package's API is pretty small, and I don't expect it to grow considerably over its lifetime. So perhaps it is better overall not to introduce `func Allow(level Value) Option` and deprecate `AllowError()` and friends purely because it doesn't buy a whole lot and leaves the API looking a bit dingy in the end.

So I am OK with adding a `level.???(userString, default)` type API.

As for naming, I think we should stay consistent with the existing patterns in the package. None of the existing functions that return an `Option` have "Option" in the name, so let's not do that. All of the existing options that control the allowed levels start with "Allow" so let's embrace that. Therefore I propose:

`func AllowParsed(level string, default Option) Option`

My second choice is `AllowFromString` with the same signature. It's my second choice because it feels less refined to me.

Reply:
> `func AllowParsed(level string, default Option) Option`
> `func AllowFromString(level string, default Option) Option`

I like either one.

Reply:
> > `func AllowParsed(level string, default Option) Option`
> > `func AllowFromString(level string, default Option) Option`
> 
> I like either one.

ok, my first commit was with a method called `AllowByString` so quite close. 😄 

I'd really like to finalize this PR. can we find a final agreement on:
- which API to implement: just one function or a set of variants like
  - level.???(level string) (Option, error)
  - level.Must???(level string). <-- panic if parse fails
  - level.???WithDefault(level string, def Option) Option
- naming

Reply:
> > `func AllowParsed(level string, default Option) Option`
> > `func AllowFromString(level string, default Option) Option`
> 
> I like either one.

Me too.

Reply:
looks like we agreed on a single function 
```go
func AllowParsed(level string) (Option, error)
```

May I proceed with the implementation ?

Reply:
> I am not convinced about a version that accepts a default. 

I think it's important to enable the common use case — decorating a logger with a level filter based on user input, with a reasonable default — with a single expression. That is, I specifically want to avoid any conditionals in user code. I don't see how to do that without either panic or some kind of default.

Reply:
I understand that desire. I don't think it is as important as you do, but I'm willing to compromise. I don't like the function with a default fallback because it enables misconfigurations to persist unidentified. If we must add a function that returns a single `Option` for use inline, then I would rather add a function that panics alongside the one that returns an error since the panic is something the caller can opt into if they want the convenience.

In that scenario we would be adding

```
func AllowParse(level string) (Option, error)
// I dropped the final 'd' in the name here to align with the next function which would be weird if it had a final 'd'.

func AllowMustParse(level string) Option
```

Is that acceptable?

Reply:
What about
```go
func AllowFromString(level string)  (Option, error)

func MustAllowFromString(level string)
```
?

Reply:
In `MustAllowFromString` I think the "must" is modifying the wrong thing from an english grammar perspective. It will also prevent the function from sorting next to the other "Allow..." functions in the package docs.

Reply:
mmmh, well I was suggesting it because something similar is implemented by some packages (for instance `jmoiron/sqlx`) where you usually have
```go
func ???(args) (R, error)

func Must???(args) R
```

and the `Must???` panics in case of error.

I'm anyway open to use the names the repository's maintainers decide.

Reply:
Yeah, I know the convention you are thinking of, the standard library does that as well. I think the names would be cleaner if we had standalone parsing functions as we discussed above, but we seem to be split on that idea. When we have a function that parses and and returns an `Option` it makes the naming trickier. Since the function returns an `Option` I think it should follow the naming pattern of the other functions that return an `Option`.

<img width="377" alt="Screen Shot 2022-04-08 at 10 00 16 AM" src="https://user-images.githubusercontent.com/1060711/162451593-74ff0b7b-2fea-40ec-929a-d15e52d60fec.png">

The parsing behavior is a secondary means to an end. That's why I constructed the names I suggested the way I did.

If the name starts with "Must" I think it obscures the primary goal of the function and, as I said above, it will not appear as well grouped in the list of options in the package docs.

Reply:
Thanks, @ChrisHines, for that reminder about docs. It make me realize that the proposed parsing function would be the only Option constructor that returned `(Option, error)` instead of just `Option`, and the fact that this would "stand out" in this way feels like a negative signal for the design. 

Going back to @sagikazarmark's original intuition...

```go
func Allow(v Value) Option

func ParseLevel(s string) (Value, error)

func ParseLevelDefault(s string, def Value) Value
```

which would allow

```go
var logger log.Logger
logger = log.NewLogfmtLogger(os.Stdout)
logger = level.NewFilter(logger, level.Allow(level.ParseLevelDefault(s, level.Info)))
```

?

Reply:
@peterbourgon, I'm glad the visual helped. I think you're now in alignment with what @sagikazarmark and I were advocating for above. I like your functions, with one nit. `level.ParseLevel` and `level.ParseLevelDefault` stutter a bit. Should they be `level.Parse` and `level.ParseDefault`?

Reply:
@ChrisHines 

> `level.ParseLevel` and `level.ParseLevelDefault` stutter a bit. Should they be `level.Parse and `level.ParseDefault`?

They should! I thought you expressed some dissatisfaction with these names, but if not, then 100% on board.

Reply:
🎉 Great. 🎉  I'm on board too.